### ISSUE TITLE
Refactor to have a FixedEffects class

### ISSUE THREAD

**Issue by notcome** [xuyiqing/fastplm#9]
Currently, we have a class `FixedEffect` which:

* keeps the indicator, number of levels, and size of each level,
* converts the indicator from R to C++ (i.e. one-based numbering to zero-based numbering),
* has an accelerated `computeMean` that computes the mean all at once,
* has a method `demean` that demeans via a queue that might execute in parallel.

Issues:

* We cannot extend the model to contain interacting effects.
* We cannot integrate component analysis without duplicating code.

Also, we use a C-style function for the parallel computing to avoid call overhead. Even calling a class method would double the execution time, not to mention `std::function`. However, a better solution is to move the MAP convergence check to `demean` altogether and increase the granularity of execution blocks.

Proposal: add a `FixedEffects` class to manage all fixed effects. There should be a clear boundary between C++ side and R side. In particular, `rowname` and `colname` should stay in R.

I did not follow this path because such a class would mutate throughout its lifecycle. The irony is that all my other classes mutate anyway. To resolve the issue, we should follow the new pattern used in the new code:
1. Use a static factory method to do the indicator-converting and component analysis.
2. Demean should not mutate (i.e. `deltas` should not be a property but a pass-in references).
3. When checking cross-component interaction, there should be no mutation.
In other words, the factory method should return a frozen object.

Step-by-step proposal:
1. Integrate the current code into the new class without adding any new functionality. In particular, transfer the convergence check to `FixedEffects`.
2. Add support component analysis.
3. Add support for interacting effects.

From now on, we will also use GitHub issues to track commits. I will open an new issue for each step.
